server.c

```diff
@@ -0,0 +1,292 @@
+    set->data = calloc(set->capacity, sizeof(struct node_t));
+}
+
+void
+set_insert(struct set_t *set, struct node_t node)
+{
+    set->data[set->size] = node;
+    set->size++;
+    if (set->size == set->capacity)
+    {
+        set->capacity *= 2;
+        set->data = realloc(set->data, set->capacity);
+    }
+}
+
+// More convenient than set_insert in this case
+struct node_t *
+set_take_last(struct set_t *set)
+{
+    set->size++;
+    if (set->size == set->capacity)
+    {
+        set->capacity *= 2;
+        set->data = realloc(set->data, set->capacity);
+    }
+    return &set->data[set->size - 1];
+}
+
+void
+set_remove_sock(struct set_t *set, int socket_fd)
+{
+    int idx = 0;
+    for (int i = 0; i < set->size; ++i)
+    {
+        if (set->data[i].socket_fd == socket_fd)
+        {
+            idx = i;
+            break;
+        }
+    }
+    set->data[idx] = set->data[set->size - 1];
+    set->size--;
+}
+
+void
+set_destroy(struct set_t *set)
+{
+    free(set->data);
+}
+
+struct srv_context_t
+{
+    volatile int tasks_running;
+    pthread_mutex_t tasks_mutex;
+    pthread_cond_t tasks_cond;
+
+    struct set_t set;
+    struct queue_t queue;
+    password_t password;
+    char *hash;
+    volatile bool found;
+    volatile bool done;
+
+    struct config_t *config;
+};
+
+struct params_t
+{
+    struct srv_context_t *context;
+    int socket_fd;
+};
+
+static int
+send_task(const int client_sfd, struct task_t *task, bool *result)
+{
+    int status;
+
+    status = send(client_sfd, (char *) task, sizeof(struct task_t), 0);
+    if (status <= 0) return -1;
+
+    int size;
+    status = recv(client_sfd, &size, sizeof(size), 0);
+    if (status <= 0) return -1;
+
+    if (size != 0)
+        recv(client_sfd, task->password, size, 0);
+
+    *result = (size != 0);
+    return 0;
+}
+
+static void *
+serve_client(void *arg)
+{
+    struct params_t *params = (struct params_t *) arg;
+    struct srv_context_t *context = (struct srv_context_t *) params->context;
+    int client_sfd = params->socket_fd;
+
+    while (true)
+    {
+        struct task_t task;
+        queue_pop(&context->queue, &task);
+
+        task.to = task.from;
+        task.from = 0;
+        bool found = false;
+        int status = send_task(client_sfd, &task, &found);
+        if (status == -1)
+        {
+            queue_push(&context->queue, &task);
+            set_remove_sock(&context->set, client_sfd);
+            close(client_sfd);
+            break;
+        }
+        if (found)
+        {
+            memcpy(context->password, task.password, sizeof(task.password));
+            context->found = true;
+            context->done = true;
+        }
+
+        pthread_mutex_lock(&context->tasks_mutex);
+        --context->tasks_running;
+        pthread_mutex_unlock(&context->tasks_mutex);
+
+        if (context->tasks_running == 0 || context->found)
+            pthread_cond_signal(&context->tasks_cond);
+    }
+
+    return NULL;
```

Comment:
Exit from client thread should remove client thread id from the set of running threads.
It means that set should have a lock on it.